### ISSUE TITLE
feature/mobx@next warning at /counter url

### ISSUE THREAD

**Issue by saniko** [ctrlplusb/react-universally#404]
When opening the app at http://localhost:1337/counter
Any button click throws this warning:

`Warning: forceUpdate(...): Can only update a mounted or mounting component. This usually means you called forceUpdate() on an unmounted component. This is a no-op. Please check the code for the Counter component.`

FYI

**Comment by ctrlplusb** [comment]
Will be interesting to see if the latest merge of `next` into the mobx branch solves this.

**Comment by birkir** [comment]
Still an issue 😞 

After investigating this issue better, I came to the conclusion that all packages that piggyback the `componentWillMount` method of an class will be frown to this issue (lots of packages man...).

Simple test:
 - Add `this.forceUpdate();` into your `componentWillMount` and see what the fuzz is all about.
 - Also you can add a `console.log` there and you will see it fire twice on the server, twice on the client.

**Comment by ctrlplusb** [comment]
Is this only an issue when used in conjunction with `react-async-component`?

**Comment by diondirza** [comment]
@ctrlplusb Just running fresh pull from next branch, seems navigating to another link produce page refresh. Got an error from console related to `asyncComponent`.

<img width="690" alt="screen shot 2017-04-05 at 5 45 37 pm" src="https://cloud.githubusercontent.com/assets/10954870/24702151/b98b4442-1a27-11e7-878b-4175804de4de.png">

**Comment by birkir** [comment]
@diondirza  The next branch has been updated with new react-async-component so this is irrelevant error.

@ctrlplusb Yes, this only happens with react-async-component, but it does not matter if you lazy load a component or not, react-tree-walker will always fire componentWillMount twice. And this is just not to the spec of React.

**Comment by ctrlplusb** [comment]
Hey @birkir thanks for this. It definitely should not execute componentWillMount twice within react-tree-walker. That would be a big. Have you tried latest? I updated the react-tree-walker API recently.

We should only have two componentWillMoint hits. Once within react-tree-walker and then once within the render/renderToString

**Comment by ctrlplusb** [comment]
I am gonna pull down your latest mobx and investigate.

**Comment by ctrlplusb** [comment]
Busy updating the branch to latest

**Comment by ctrlplusb** [comment]
Updated to latest version of `next` which includes new version of `react-async-component`.  

Works, no errors/warnings. 🎉

**Comment by birkir** [comment]
Wait, this is not really fixed, at least not the forceUpdate thing, maybe something else fixed?

A `this.forceUpdate()` inside componentWillMount will still trigger the error.

Also thinking... why would you want to fire componentWillMount twice. This is a lifecycle method in react components that is often used to do async work with the component.

Checkout this observer mixin for classes that hooks into componentWillMount.
https://github.com/mobxjs/mobx-react/blob/master/src/observer.js#L106-L232

**Comment by ctrlplusb** [comment]
Hmmm, you shouldn't be firing this.forceUpdate on the server though.  I checked that source and they look for a `__$mobxIsUnmounted` which gets set via the `componentWillUnmount`. But on the server that lifecycle event should never get call.  So perhaps this is an issue with mobx expectations vs ssr react expectations?

**Comment by ctrlplusb** [comment]
I need to play with mobx to get some more insight into this.

**Comment by ctrlplusb** [comment]
Some more details here: https://github.com/mobxjs/mobx-react/issues/12

Seems to be a known issue with mobx.  Even without react-async-component, calling renderToStaticMarkup/String can easily finish before an async mobx event completes, and the same would occur.

**Comment by birkir** [comment]
Oh okay, sorry to bother you with this as this seems to be non-project related issue.
Thanks for your time.

**Comment by ctrlplusb** [comment]
No worries, this is good for me to be aware of. I'd like to try spend some more time on it to be honest and hopefully come up with something more helpful. Mobx is pretty big so having full ssr support would be epic.

**Comment by xiao-hu** [comment]
I am using `react-async-component` with `mobx` and I encountered exactly the same error of `forceUpdate(...)` (on client side). Saw this post and I upgraded `react-async-component` to the latest version 1.0.0-beta.2 (`mobx` 3.1.8). But the same error is still there. And `componentWillMount` is still called twice on client and twice on server. Did I miss any thing?

![image](https://cloud.githubusercontent.com/assets/6172805/24732939/47576d36-1a42-11e7-80ad-071a9612f075.png)
